opm/ml/keras_model.cpp

```diff
@@ -0,0 +1,419 @@
+    KASSERT(ReadUnsignedInt(file, &activation),
+            "Failed to read activation type");
+
+    switch (activation) {
+    case kLinear:
+        activation_type_ = kLinear;
+        break;
+    case kRelu:
+        activation_type_ = kRelu;
+        break;
+    case kSoftPlus:
+        activation_type_ = kSoftPlus;
+        break;
+    case kHardSigmoid:
+        activation_type_ = kHardSigmoid;
+        break;
+    case kSigmoid:
+        activation_type_ = kSigmoid;
+        break;
+    case kTanh:
+        activation_type_ = kTanh;
+        break;
+    default:
+        KASSERT(false, "Unsupported activation type %d", activation);
+    }
+
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerActivation<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    *out = *in;
+
+    switch (activation_type_) {
+    case kLinear:
+        break;
+    case kRelu:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            if (out->data_[i] < 0.0) {
+                out->data_[i] = 0.0;
+            }
+        }
+        break;
+    case kSoftPlus:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            out->data_[i] = log(1.0 + exp(out->data_[i]));
+        }
+        break;
+    case kHardSigmoid:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            Evaluation x = (out->data_[i] * 0.2) + 0.5;
+
+            if (x <= 0) {
+                out->data_[i] = 0.0;
+            } else if (x >= 1) {
+                out->data_[i] = 1.0;
+            } else {
+                out->data_[i] = x;
+            }
+        }
+        break;
+    case kSigmoid:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            Evaluation& x = out->data_[i];
+
+            if (x >= 0) {
+                out->data_[i] = 1.0 / (1.0 + exp(-x));
+            } else {
+                Evaluation z = exp(x);
+                out->data_[i] = z / (1.0 + z);
+            }
+        }
+        break;
+    case kTanh:
+        for (size_t i = 0; i < out->data_.size(); i++) {
+            out->data_[i] = sinh(out->data_[i])/cosh(out->data_[i]);
+        }
+        break;
+    default:
+        break;
+    }
+
+    return true;
+}
+
+
+template<class Evaluation>
+bool KerasLayerScaling<Evaluation>::LoadLayer(std::ifstream* file) {
+    KASSERT(file, "Invalid file stream");
+
+    KASSERT(ReadFloat(file, &data_min), "Failed to read min");
+    KASSERT(ReadFloat(file, &data_max), "Failed to read max");
+    KASSERT(ReadFloat(file, &feat_inf), "Failed to read max");
+    KASSERT(ReadFloat(file, &feat_sup), "Failed to read max");
+    return true;
+}
+
+template<class Evaluation>
+bool KerasLayerScaling<Evaluation>::Apply(Tensor<Evaluation>* in, Tensor<Evaluation>* out) {
+    KASSERT(in, "Invalid input");
+    KASSERT(out, "Invalid output");
+
+    Tensor<Evaluation> temp_in, temp_out;
+
+    *out = *in;
```

Comment:
Do you require that out is the same size as in already? If so, you can skip this copy assignment and just use `in.data_[i]` in line 186. Otherwise, a simple resize here would do the trick. Same applies to other classes here.

You may want to add a resize() member to Tensor that takes a std::vector<int> as argument instead of just the 1, 2, 3, 4 dimensional current versions.